### DISCUSSION TITLE
How do I ensure supplementary files are present on quarto.pub?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#6487] [Q&A]
### Description

At JSM, there was an attempt to use <https://quarto.pub> to host a quarto document that used an extension needing two additional files to be placed alongside the document. From the error message below, it seems like the files were not included alongside the rendered document.

Is there a way to ensure files are included with rendered quarto documents on quarto.pub?

Screenshot of the page: 

<img width="1301" alt="Screenshot 2023-08-10 at 11 43 54 AM" src="https://github.com/quarto-dev/quarto-cli/assets/833642/5175861a-e412-459e-828b-31de7b5f89cb">

Relevant error message from the demo showing missing files:

```
maria.quarto.pub/:1 Uncaught (in promise) TypeError: Failed to register a ServiceWorker for scope ('https://maria.quarto.pub/webr-demo/') with script ('https://maria.quarto.pub/webr-demo/webr-serviceworker.js'): A bad HTTP response code (404) was received when fetching the script.
```

Demo page on quarto.pub:

<https://maria.quarto.pub/webr-demo/>

Local version working:

<https://www.youtube.com/watch?v=Kb_5GM3aUSk>

**Comment by dragonstyle** [comment]
Typically, we expect the extension author to provide these dependencies to Quarto so they may be moved alongside the rendered output. This is usually done via the LUA API:

https://quarto.org/docs/extensions/lua-api.html#html-dependencies

It looks like perhaps the service worker needs to be provided (but you may provide additional scripts or other types of HTML dependencies as well).

**Reply by mcanouil** [reply]
FYI, but not addressing directly the question:

- <https://quarto.org/docs/websites/website-tools.html#site-resources>

**Reply by coatless** [reply]
Alrighty, I'll give that a try. Before, I seem to be able to get away with using `quarto.doc.add_format_resource()`. Does this indicate that `quarto.pub` is preventing the `.js` output?

On another note, it looks like `serviceworkers` was slightly misspelled though. So, I've opened a PR:  <https://github.com/quarto-dev/quarto-cli/pull/6490>

Lastly, for my own sanity and a perhaps an additional documentation example, I think adding serviceworkers with custom names would be given by:

```lua
  quarto.doc.add_html_dependency({
    name = "webr-serviceworker",
    version = "0.1.1",
    serviceworkers = {path = "webr-serviceworker.js", name = "remapped-webr-serviceworker.js"} 
  })
```

For just a regular string, all that is required would be:

```lua
  quarto.doc.add_html_dependency({
    name = "webr-serviceworker",
    version = "0.1.1",
    serviceworkers = {"webr-serviceworker.js"} 
  })
```

**Comment by dragonstyle** [comment]
`quarto.doc.add_format_resource()` will copy the file from your extension to the root directory (so if you emit the correct JS and add a format resource, that may work as long as we 'see' the dependency, which I _think_ we would?). I'm not sure why that would've changed, honestly, but it does seem like we must not be discovery that dependency when moving files to the output directory for publishing.

Your examples above are exactly correct!

**Reply by coatless** [reply]
So, I've tried to publish a demo Qmd file onto Quarto pub with the updated structure now; but, even with the above changes (using `quarto.doc.add_html_dependency()`), I'm getting a similar error on page load.

<https://coatless.quarto.pub/webr-enabled-code-cells/>

Is there a way to see what files are being uploaded? Or what files are present with a quarto.pub site?

I published to <https://quarto.pub> using the publishing command with `QUARTO_PUB_AUTH_TOKEN`.

In particular, we have:

```sh
@coatless ➜ /workspaces/quarto-webr (main) $ quarto publish quarto-pub webr-demo.qmd
```

Gives:
 
```sh
? Publish with account: › QUARTO_PUB_AUTH_TOKEN
? Document name: › webr-enabled-code-cells

[✓] Creating quarto-pub document

Rendering for publish:

processing file: webr-demo.qmd
1/5                  
2/5 [unnamed-chunk-1]
3/5                  
4/5 [unnamed-chunk-2]
5/5                  
output file: webr-demo.knit.md

Warning messages:
1: In grSoftVersion() :
  unable to load shared object '/usr/local/lib/R/modules//R_X11.so':
  libXt.so.6: cannot open shared object file: No such file or directory
2: In get_engine(options$engine) :
  Unknown language engine 'webr' (must be registered via knit_engines$set()).
pandoc 
  to: html
  output-file: webr-demo.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  toc: true
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: webR-enabled code cells
  
[✓] Preparing to publish document
[✓] Uploading files (complete)
[✓] Deploying published document
[✓] Published document: https://coatless.quarto.pub/webr-enabled-code-cells
[✓] Account site updated: https://coatless.quarto.pub
```

This was built and published using `quarto-1.4.330` and `quarto-webr` @ v0.3.0
